fix: https passthrough. aiohttp_client compatability

brycedrennan opened the pull request:
This is an alternative to #55 that maintains compatibility with pytest-asyncio and fixes #51 and fixes #52.

The `is_ssl` method was monkey-patched in a synchronous way that couldn't work when simultaneous requests needed it both patched and not-patched.  Instead of undoing the ClientRequest monkey-patching we create DirectClientRequest and have our passthrough session use it instead.

For compatibility between aresponses and aiohttp_client fixtures we use a work-around recommended by @serjshevchenko  (https://github.com/pytest-dev/pytest-asyncio/issues/170#issuecomment-706114516) that replaces the aiohttp_client loop fixture with the event loop provided by pytest-asycnio.  The hanging was caused by different fixtures running on different event loops.

Workaround example:
```python
@pytest.fixture
def loop(event_loop):
    """replace aiohttp loop fixture with pytest-asyncio fixture"""
    return event_loop
```

Also added `add_local_passthrough` which is useful when working with your own application server or other local resources.  To work with `assert_plan_strictly_followed` you should pass in the number of times it will be called as the `repeat` argument.

It's equivalent to
```python
LOCALHOST = re.compile(r"127\.0\.0\.1:?\d{0,5}")
aresponses.add(
    host_pattern=LOCALHOST,
    repeat=aresponses.INFINITY,
    response=aresponses.passthrough
    )
```

@elupus @Zopieux I think I prefer this solution over #55 although it does require the developer to manually override the `loop` fixture.  I'll leave these pull requests open for a few days to give you both an opportunity to comment.

Zopieux commented on README.md:
````diff
@@ -213,6 +214,15 @@ async def aresponses(loop):
         yield server
 ```
 
+If you're trying to use the `aiohttp_client` test fixture then you'll need to mock out the aiohttp `loop` fixture
+instead:
+```python
+@pytest.fixture
+def loop(event_loop):
+    """replace aiohttp loop fixture with pytest-asyncio fixture"""
+    return event_loop
+```
+
````
More importantly, this needs to go in a proper documentation page. I'd not consider a Git/GitHub README to be part of the things you *have* to read to use a library.

brycedrennan replied:
For now, the readme is our only documentation.  Changing that would be outside the scope of this PR.

Zopieux replied:
Ah, lol I did not even realize that until now sorry.

Zopieux commented on requirements.in:
```diff
@@ -1,10 +1,9 @@
-pytest==5.3.1
-pytest-asyncio==0.10.0
-pytest-cov==2.8.1
+coverage
+pydocstyle
+pylava
+pylava-pylint
+pylint
+pytest
+pytest-cov
```
I'm not this project's author but this change seems unwise. `requirements.txt` is there to have a reproducible dev environment, so version pinning is key. Use setup `install_requires` for providing dependency version ranges  (with proper CI testing coverage for hassle-free/automated version bumps).

I'd also consider this to be outside of the scope of the PR (not atomic).

brycedrennan replied:
This is not a requirements.txt file, that's still there and everything is pinned. 

I agree it would be better in a different PR. There were dependency conflicts and this was the fastest way to resolve them.

brycedrennan replied:
I have moved these changes outside of this PR. much cleaner now.

brycedrennan commented:
@Zopieux thanks for taking a look at this.

Zopieux reviewed (commented):
Please keep the PR atomic by removing unrelated changes (eg. Makefile, pinning, tox.ini).

Zopieux commented:
> The hanging was caused by different fixtures running on different event loops.

\*sigh\* Thanks for looking. That makes sense, and this always has been a huge hassle in asyncio libraries and projects. This is often a friction point when it comes to making two libraries interact, all the more with magic-heavy things such as pytest.

Given that pytest-aiohttp is 1. popular 2. [documented as the go-to solution](https://docs.aiohttp.org/en/v3.7.3/testing.html#testing-aiohttp-web-servers), I'd obviously prefer a solution that is not incompatible with it, even if it requires some extra setup work. At the same time, pytest-aiohttp looks kinda abandonware at this point: https://github.com/aio-libs/pytest-aiohttp/issues/13, https://github.com/aio-libs/pytest-aiohttp/issues/16.

brycedrennan commented:
@Zopieux I've addressed your feedback and I think you're saying you are okay with this approach. Let me know if you have anything further.

FYI pytest-aiohttp is just a pointer to the testing fixture built in to aiohttp itself, so active development would be happening there. It makes sense that there wouldn't be much activity in the actual pytest-aiohttp repo.

elupus commented:
If this works.. But those two pytest add-ons are inherently incompatible.

I don't see why this lib should depend on either. I agree it is needed for its own test suite. But it should be possible to make sure the lib when used by end users ie standard pip install case does not bring in any of those test frameworks.

elupus commented:
See for example here on how to only add that dependency for tests: https://github.com/elupus/nibeuplink/blob/ea12d4b63c5cb23923d0d970c2d992356b83120a/setup.py#L37

brycedrennan commented:
> I don't see why this lib should depend on either. 

The aresponses fixture has to depend on an event loop.  See https://github.com/CircleUp/aresponses/blob/master/aresponses/main.py#L266 

Since pytest-asyncio covers a wider set of use cases I went with that one.  If there is some other option let me know.

elupus commented:
Yes. But that is an extremely simple fixture. It could easily introspect to see which pytest addon is available and just add a fixture that is compatible. 

Or it could just be part of documentation.

brycedrennan commented:
> It could easily introspect to see which pytest addon is available and just add a fixture that is compatible.

I don't know how to do this, how would I do this?

Zopieux commented:
Careful with introspection code that tries to bring compatibility with various versions of various similar libraries. This is doomed to age badly and break. I'd much prefer documentation with 2-line code stubs for each poplar library. This allows developers to understand and maintain the compatibility "layer" in their own code, explicitly.

As long as that layer is thin and documented, which in this case it is (right?), I'd prefer this over tech debt magic. 👍
